WIP: Fiscal Emitter

igorocampos opened the pull request:
fixes #95

@dicksonkimeu I've started this PR with the base for all commands, You can check here what files I added/changed, but I think the best before continuing the implementation is to pull from this branch (`fiscal`) , test and tweak with it while debugging with your fiscal printer until the commands are working. I have implemented paper feed, cut and Non-Fiscal Receipt printing. 

Please check it out, and reach out for any question

[DATECS FP-300.pdf](https://github.com/lukevp/ESC-POS-.NET/files/5832381/DATECS.FP-300.pdf)

dicksonkimeu commented:
Hello.

It has responded well on paper feed. We can do fiscal receipt then I do
complete test.

We can do receipt.

On Thu, 24 Dec 2020, 06:25 Igor, <notifications@github.com> wrote:

> fixes #95 <https://github.com/lukevp/ESC-POS-.NET/issues/95>
>
> @dicksonkimeu <https://github.com/dicksonkimeu> I've started this PR with
> the base for all commands, You can check here what files I added/changed,
> but I think the best before continuing the implementation is to pull from
> this branch (fiscal) and test in your fiscal printer if the commands are
> working. I have implemented paper feed, cut and Non-Fiscal Receipt printing.
>
> Please check it out, and reach out for any question
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/lukevp/ESC-POS-.NET/pull/99
> Commit Summary
>
>    - Merge pull request #3 from igorocampos/qrcode_length
>    - Merge pull request #2 from igorocampos/unit_tests
>    - Merge pull request #4 from lukevp/master
>    - Added structure for Fiscal Printer implementation
>    - Implemented Paper Cut, Paper Feed and Non Fiscal Receipt
>
> File Changes
>
>    - *M* ESCPOS_NET.ConsoleTest/Program.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-d82854e0f89945cba1b24ea41318889aa54a344f9a8bb85249a6b138e47cc69c>
>    (92)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/ActionCommands.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-dfe6c1c108739ba5aeb9cdac9cfb5bc414d3e8235f731e893959ffef6a95b105>
>    (11)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/BaseCommands.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-b63eff6389dc38e6a8b50915931b7707f3d2758a110f40acbd641ce16cfee5c9>
>    (60)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/NonFiscalCommands.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-0efde4707ed79a37b71227198a07bf38ecc9c5e0aabef96777f11101a8677c67>
>    (43)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/Cmd.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-648ad67710bad714ad5b6b6a99560a7c7091840ca446c5ed51787185ae1b2cc1>
>    (10)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/Functions.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-58e2b102ae43a58ad5ce69ac3fbd9facf76f629b13d499460f411abbd914d324>
>    (8)
>    - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/NonFiscal.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-da1419bfe06d2b74531f7f69443c812ef3857039890789d191d25945a9e334af>
>    (9)
>    - *A* ESCPOS_NET/FiscalEmitters/DATECS.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-dc5aa5871239cc03585c8108261b07caf6aa5f9a936e51072a41c75478bdecd4>
>    (6)
>    - *A* ESCPOS_NET/FiscalEmitters/ICommandEmitter.cs
>    <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-e95a5804be187abc1611a1fd79edc9ffc150130c66826306ad10468419fdcf3f>
>    (10)
>
> Patch Links:
>
>    - https://github.com/lukevp/ESC-POS-.NET/pull/99.patch
>    - https://github.com/lukevp/ESC-POS-.NET/pull/99.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lukevp/ESC-POS-.NET/pull/99>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCE7FRLPRT5ST6OMFMRG33SWKYDFANCNFSM4VHW7NWQ>
> .
>

igorocampos commented:
> Hello. It has responded well on paper feed. We can do fiscal receipt then I do complete test. We can do receipt.
> […](#)
> On Thu, 24 Dec 2020, 06:25 Igor, ***@***.***> wrote: fixes #95 <#95> @dicksonkimeu <https://github.com/dicksonkimeu> I've started this PR with the base for all commands, You can check here what files I added/changed, but I think the best before continuing the implementation is to pull from this branch (fiscal) and test in your fiscal printer if the commands are working. I have implemented paper feed, cut and Non-Fiscal Receipt printing. Please check it out, and reach out for any question ------------------------------ You can view, comment on, or merge this pull request online at: #99 Commit Summary - Merge pull request #3 from igorocampos/qrcode_length - Merge pull request #2 from igorocampos/unit_tests - Merge pull request #4 from lukevp/master - Added structure for Fiscal Printer implementation - Implemented Paper Cut, Paper Feed and Non Fiscal Receipt File Changes - *M* ESCPOS_NET.ConsoleTest/Program.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-d82854e0f89945cba1b24ea41318889aa54a344f9a8bb85249a6b138e47cc69c> (92) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/ActionCommands.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-dfe6c1c108739ba5aeb9cdac9cfb5bc414d3e8235f731e893959ffef6a95b105> (11) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/BaseCommands.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-b63eff6389dc38e6a8b50915931b7707f3d2758a110f40acbd641ce16cfee5c9> (60) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandEmitter/NonFiscalCommands.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-0efde4707ed79a37b71227198a07bf38ecc9c5e0aabef96777f11101a8677c67> (43) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/Cmd.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-648ad67710bad714ad5b6b6a99560a7c7091840ca446c5ed51787185ae1b2cc1> (10) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/Functions.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-58e2b102ae43a58ad5ce69ac3fbd9facf76f629b13d499460f411abbd914d324> (8) - *A* ESCPOS_NET/FiscalEmitters/BaseCommandValues/NonFiscal.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-da1419bfe06d2b74531f7f69443c812ef3857039890789d191d25945a9e334af> (9) - *A* ESCPOS_NET/FiscalEmitters/DATECS.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-dc5aa5871239cc03585c8108261b07caf6aa5f9a936e51072a41c75478bdecd4> (6) - *A* ESCPOS_NET/FiscalEmitters/ICommandEmitter.cs <https://github.com/lukevp/ESC-POS-.NET/pull/99/files#diff-e95a5804be187abc1611a1fd79edc9ffc150130c66826306ad10468419fdcf3f> (10) Patch Links: - https://github.com/lukevp/ESC-POS-.NET/pull/99.patch - https://github.com/lukevp/ESC-POS-.NET/pull/99.diff — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#99>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABCE7FRLPRT5ST6OMFMRG33SWKYDFANCNFSM4VHW7NWQ> .

How about the other the non fiscal receipt? That is also implemented as well.
It should open a non fiscal receipt, print a text and close the non fiscal receipt.

dicksonkimeu commented:
Hi, let me test and get back. On 28/29th.

dicksonkimeu commented:
Hi tested. It works well. We can now add a Fiscal Receipt.

igorocampos commented:
Hey @lukevp Merry Christmas! 

Let me ask you something: I was implementing some fiscal commands that have data validation according to DATECS handbook, but it is clearly something that is not going to be the same for all manufacturers. So I would like to provide a DataValidator that could act differently for each FiscalEmitter instance. This is a little bit different from what you did with the DataValidation for plain Emitters because in there you will always call the same validator, and I wish to have multiple validators for the same method where each manufacturer can have its own. For example:
- There is a command that registers one item to the fiscal receipt at the sale. All manufacturers have this command, however the data validation can be very different, for instance, DATECS do now allow more than 3 decimal places for Quantity, nor Discount above 99.00%. But I've worked with other fiscal printers that allowed 4 decimal places and 99.99% discount.

Anyways, please take a look at the commit I'm about to do now, and see if you would like a different architecture or not.

igorocampos commented:
@dicksonkimeu I've just pushed a commit for fiscal receipt usage. Please test it out. You must:
- Open the receipt
- Register 1 or more items
- Register 1 or more totalizators
- Print (or not) any text
- Close the Receipt

dicksonkimeu commented:
@igorocampos I will test on Monday and revert..

dicksonkimeu commented:
@igorocampos am preparing my code. To ensure I don't miss something any sample code ?.

igorocampos commented:
> @igorocampos am preparing my code. To ensure I don't miss something any sample code ?.

Just use the same way you tested the paper feed and non fiscal receipt.

You can also look in this PR the exact name of the command methods

dicksonkimeu commented:
Have run tests, doesn't return any error and no response in printer

```
 try
            {
                List<Item> items = new List<Item>();
                Item item1 = new Item();
                item1.Name = "10001";
                item1.ProductID = "Afia Juice 900 ml";
                item1.Quantity = 4;
                item1.UnitPrice = 30;
                string Descriptopn1 = "10293030\n";
                Descriptopn1 = Descriptopn1 + "" + "Fresh Juice 500 ML";
                item1.Descriptopn = Descriptopn1;

                Item item2 = new Item();
                item2.Name = "10002";
                item2.ProductID = "Fresh Juice 3900 ML";
                item2.Quantity = 20;
                item2.UnitPrice = 10;
                string Descriptopn2 = "10293030\n";
                Descriptopn2 = Descriptopn2 + "" + "Fresh Juice 500 ML";
                item2.Descriptopn = Descriptopn2;
                

                items.Add(item1);
                items.Add(item2);

                var e1 = new DATECS();
                var printer = new SerialPrinter(portName: "COM9", baudRate: 115200);
                printer.Write(e1.OpenReceipt());
                foreach (Item item in items)
                {
                    printer.Write(e1.RegisterItem(item.Descriptopn, char.Parse("A"), item.UnitPrice, item.Quantity));
                }
                printer.Write(e1.RegisterTotal("CASH", char.Parse("P"), 320));
                printer.Write(e1.PrintText("Thank you for shopping"));
                printer.Write(e1.CloseReceipt());
            }
            catch (Exception ex)
            {
                string error = ex.Message;
            }

```

dicksonkimeu commented:
Have run tests, doesn't return any error and no response in printer

```
 try
            {
                List<Item> items = new List<Item>();
                Item item1 = new Item();
                item1.Name = "10001";
                item1.ProductID = "Afia Juice 900 ml";
                item1.Quantity = 4;
                item1.UnitPrice = 30;
                string Descriptopn1 = "10293030\n";
                Descriptopn1 = Descriptopn1 + "" + "Fresh Juice 500 ML";
                item1.Descriptopn = Descriptopn1;

                Item item2 = new Item();
                item2.Name = "10002";
                item2.ProductID = "Fresh Juice 3900 ML";
                item2.Quantity = 20;
                item2.UnitPrice = 10;
                string Descriptopn2 = "10293030\n";
                Descriptopn2 = Descriptopn2 + "" + "Fresh Juice 500 ML";
                item2.Descriptopn = Descriptopn2;
                

                items.Add(item1);
                items.Add(item2);

                var e1 = new DATECS();
                var printer = new SerialPrinter(portName: "COM9", baudRate: 115200);
                printer.Write(e1.OpenReceipt());
                foreach (Item item in items)
                {
                    printer.Write(e1.RegisterItem(item.Descriptopn, char.Parse("A"), item.UnitPrice, item.Quantity));
                }
                printer.Write(e1.RegisterTotal("CASH", char.Parse("P"), 320));
                printer.Write(e1.PrintText("Thank you for shopping"));
                printer.Write(e1.CloseReceipt());
            }
            catch (Exception ex)
            {
                string error = ex.Message;
            }

```

lukevp commented:
hi @igorocampos @dicksonkimeu  sorry for the delay in replying, it has been a busy few weeks for me!  It sounds like this PR is still under test?  if so, I will not merge since I don't have a fiscalprinter to test, but I will review and provide comments shortly.

dicksonkimeu commented:
@lukevp it didn't work. i have fiscal printer fp300 datecs

igorocampos commented:
@dicksonkimeu sounds like you are not going forward into fixing what is wrong, since I don't have the fiscal printer either, I'll go ahead and close this PR. It will exist in closed PRs if ever needed for a consult.
